README.md
# Shell agent

[![Build Status](https://travis-ci.org/puppetlabs/mcollective-shell-agent.svg?branch=master)](https://travis-ci.org/puppetlabs/mcollective-shell-agent)

The shell agent allows you to start and manage shell commands via
mcollective.

It allows the running of long-running processes with a mechanism to check in
on the output from these long-running processes, which is independent of the
mcollective daemon process (the daemon can be restarted without interrupting
the processes)

To use this agent you need at least:

* MCollective 2.2.4
* Ruby 1.9 (for Process#spawn)

Please report any errors or make feature requests in the [MCOP jira project][MCOP]

Please note: we do not recommend this agent as a way of building out your
automation, for that you're still better off writing your own tailored
[agents][writing-agents] that fit your use case.  This agent is targeted
at the ad-hoc needs that people occasionally have.

[writing-agents]: http://docs.puppetlabs.com/mcollective/simplerpc/agents.html
[MCOP]: http://tickets.puppetlabs.com/browse/MCOP

## Installation

Follow the [basic plugin install guide][install guide], taking all
the code from lib and adding it to your MCollective $libdir

[install guide]: http://projects.puppetlabs.com/projects/mcollective-plugins/wiki/InstalingPlugins


## Configuring the agent

The agent should work without any additional configuration, though there are
some options you can tune the mcollective server.cfg.

### `plugin.shell.state_directory`

This is where the state used to track processes will live.  By default this
will be /var/run/mcollective-shell on Unix systems.

```
plugin.shell.state_directory = /opt/run/mcollective-shell
```


## Application usage

The `mco shell` application has several subcommands to start and manage
processes.

### mco shell run

Runs a command and reports back.  Use this for discrete short-living commands.

For long-running commands look at `start` or `run --tail`.

```
$ mco shell run dir

 * [ ============================================================> ] 2 / 2

master:
bin   dev  home  lib64       media  opt   root  selinux  srv  tmp  vagrant
boot  etc  lib   lost+found  mnt    proc  sbin  src      sys  usr  var

server2008r2a:
 Volume in drive C has no label.
 Volume Serial Number is DADF-75F9

 Directory of C:\

09/22/2012  11:45 AM    <SYMLINKD>     manifests [\\vboxsrv\manifests]
09/22/2012  11:45 AM    <SYMLINKD>     modules [\\vboxsrv\modules]
07/13/2009  08:20 PM    <DIR>          PerfLogs
09/22/2012  11:42 AM    <DIR>          Program Files
03/27/2014  06:52 AM    <DIR>          Program Files (x86)
07/03/2014  07:42 AM    <SYMLINKD>     src [\\vboxsrv\C:_src]
03/27/2014  06:39 AM    <DIR>          Users
07/03/2014  07:42 AM    <SYMLINKD>     vagrant [\\vboxsrv\vagrant]
03/27/2014  06:41 AM    <DIR>          Windows
               0 File(s)              0 bytes
               9 Dir(s)  34,565,091,328 bytes free


Finished processing 2 / 2 hosts in 221.28 ms
```

### mco shell run --tail

Starts a command, shows you the output from it, kills the command when you
interrupt with control-c, exits normally when the command exits.

```
$ mco shell -I /master/ run --tail vmstat 1

 * [ ============================================================> ] 1 / 1

master stdout: procs -----------memory---------- ---swap-- -----io---- --system-- -----cpu-----
master stdout:  r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
master stdout:  0  1 445812 120584   5808  37348   34   29    52    48   39   47  6  1 93  0  0
master stdout:  1  0 445112 122692   5824  37332 2692    0  2692    84  911 2089 47  9 40  4  0
master stdout:  1  0 444848 122576   5824  37344  288    0   288     0  773 1914 48  5 47  0  0
master stdout:  0  0 444012 121320   5824  37348 1212    0  1212     0  823 1917 47  6 45  1  0
master stdout:  0  0 443984 121204   5824  37372    0    0     0     0  797 1796 52  5 43  0  0
master stdout:  0  0 438800 117244   5824  37360 3896    0  3896     0  910 2123 49  6 45  0  0
master stdout:  1  0 438768 117136   5840  37368    0    0     0   136  811 1926 48  6 45  0  0
^CAttempting to stopping cleanly, interrupt again to kill
Sending kill to master 6dad5cb9-57f7-46e0-bad7-07ab117369a5
```


### mco shell start

Starts a command in the background and tells you the id that has been assigned
to it.  You can then use `mco shell watch`, `mco shell kill`, `mco shell list`
to monitor this process and observe its output

```
$ mco shell -I /master/ start vmstat 1

 * [ ============================================================> ] 1 / 1

master: 0dd67fac-734f-4824-8b4d-03100d4f9d07

Finished processing 1 / 1 hosts in 76.37 ms
```


### mco shell watch

Shows you the output of a command you previously started with `mco shell start`

```
$ mco shell watch 0dd67fac-734f-4824-8b4d-03100d4f9d07

 * [ ============================================================> ] 2 / 2

master stdout: procs -----------memory---------- ---swap-- -----io---- --system-- -----cpu-----
master stdout:  r  b   swpd   free   buff  cache   si   so    bi    bo   in   cs us sy id wa st
master stdout:  2  0 431448 110704   8484  40644   34   29    52    48   40   47  6  1 93  0  0
```

### mco shell list

Show a list of running jobs.

```
$ mco shell list -v

 * [ ============================================================> ] 2 / 2

master:
    0dd67fac-734f-4824-8b4d-03100d4f9d07
    1fd3961a-f48d-4119-b988-146b490a5ca3
    d174e20b-9cdb-4c14-9f34-fd29995f30cb
    ea809b20-3123-46b4-bf59-10ff7251ca9b

Finished processing 2 / 2 hosts in 142.34 ms
```

### mco shell kill

Kill a running job.

```
$ mco shell kill 0dd67fac-734f-4824-8b4d-03100d4f9d07

 * [ ============================================================> ] 2 / 2


Finished processing 2 / 2 hosts in 170.17 ms
```


Changes.md
# 0.0.2

Released 2015-03-12

* Removed saucy cows from build targets (MCOP-271)
* Add explicit `require 'pathname'` to fix agent in some configuration (PR#4)

# 0.0.1

Released 2014-07-17

* Initial public release


spec/spec_helper.rb
require 'mocha'
require 'mcollective'
require 'mcollective/test'

RSpec.configure do |config|
  config.mock_with :mocha
  config.include(MCollective::Test::Matchers)

  config.before :each do
    MCollective::PluginManager.clear
  end
end


spec/unit/mcollective/agent/shell_spec.rb
require 'spec_helper'

module MCollective
  module Agent
    describe Shell do
      let(:agent_file) { File.join('lib', 'mcollective', 'agent', 'shell.rb')}
      let(:agent) { MCollective::Test::LocalAgentTest.new('shell', :agent_file => agent_file).plugin }

      describe '#run' do
        it 'should delegate to #run_command' do
          agent.expects(:run_command).with({:command => 'echo foo'}).returns({
            :exitcode => 0,
            :stdout => "foo\n",
            :stderr => '',
          })
          result = agent.call(:run, :command => 'echo foo')
          result.should be_successful
        end
      end

      describe '#run_command' do
        let(:reply) { {} }

        before :each do
          agent.stubs(:reply).returns(reply)
          @tmpdir = Dir.mktmpdir
          Shell::Job.stubs(:state_path).returns(@tmpdir)
        end

        after :each do
          FileUtils.remove_entry_secure @tmpdir
        end

        it 'should run cleanly' do
          agent.send(:run_command, :command => 'echo foo')
          reply[:exitcode].should == 0
          reply[:stdout].should == "foo\n"
        end

        it 'should cope with large amounts of output' do
          agent.send(:run_command, :command => %{ruby -e '8000.times { puts "flirble wirble" }'})
          reply[:success].should == true
          reply[:exitcode].should == 0
          reply[:stdout].should == "flirble wirble\n" * 8000
        end

        it 'should cope with large amounts of output on both channels' do
          agent.send(:run_command, :command => %{ruby -e '8000.times { STDOUT.puts "flirble wirble"; STDERR.puts "flooble booble" }'})
          reply[:success].should == true
          reply[:exitcode].should == 0
          reply[:stdout].should == "flirble wirble\n" * 8000
          reply[:stderr].should == "flooble booble\n" * 8000
        end

        it 'raise on a non-existent command' do
          expect {
            agent.send(:run_command, :command => 'i_really_should_not_exist')
          }.to raise_error(/No such file or directory - i_really_should_not_exist/)
        end

        context 'timeout' do
          it 'should not timeout commands that exit quickly enough' do
            agent.send(:run_command, {
              :command => %{ruby -e 'puts "started"; sleep 1; puts "finished"'},
              :timeout => 2.0,
            })
            reply[:success].should == true
            reply[:exitcode].should == 0
            reply[:stdout].should == "started\nfinished\n"
            reply[:stderr].should == ''
          end

          it 'should timeout long running commands' do
            start = Time.now()
            agent.send(:run_command, {
              :command => %{ruby -e 'STDOUT.sync = true; puts "started"; sleep 5; puts "finished"'},
              :timeout => 1.0,
            })
            elapsed = Time.now() - start
            elapsed.should <= 2
            reply[:success].should == false
            reply[:exitcode].should == nil
            reply[:stdout].should == "started\n"
            reply[:stderr].should == ''
          end
        end
      end
    end
  end
end


spec/unit/mcollective/agent/shell/job_spec.rb
require 'spec_helper'
require 'mcollective/agent/shell/job'

module MCollective
  module Agent
    class Shell
      describe Job do
        before :each do
          @tmpdir = Dir.mktmpdir
          Shell::Job.stubs(:state_path).returns(@tmpdir)
        end

        after :each do
          FileUtils.remove_entry_secure @tmpdir
        end

        describe 'list' do
          it 'should return nothing when no jobs are created' do
            Job.list.should == []
          end

          it 'should return two jobs when two are made' do
            one = Job.new
            two = Job.new
            jobs = Job.list
            jobs.size.should == 2
            jobs[0].class.should == Job
            jobs[1].class.should == Job
          end
        end

        describe '#initialize' do
          it 'should use the handle if passed' do
            job = Job.new('made-up-handle')
            job.handle.should == 'made-up-handle'
          end

          it 'should generate a handle if none is specified' do
            job = Job.new
            job.handle.should =~ /[A-Z0-9-]+/
          end
        end

        describe '#start_command' do
          it 'should write files, spawn, and check for starting' do
            job = Job.new
            state_directory = job.send(:state_directory)
            job.stubs(:find_ruby).returns('testing-ruby')
            File.expects(:open).with("#{state_directory}/command", 'w')
            File.expects(:open).with("#{state_directory}/wrapper", 'w')
            Process.expects(:spawn).with('testing-ruby', "#{state_directory}/wrapper", {
              :chdir => '/',
              :in => :close,
              :out => :close,
              :err => :close,
            }).returns(53)
            File.expects(:exists?).with("#{state_directory}/pid").returns(true).twice
            IO.expects(:read).with("#{state_directory}/pid").returns("54\n")
            job.start_command('echo foo')
            job.pid.should == 54

            # explicitly unstub so the after block can fire
            File.unstub(:open)
          end
        end

        describe '#stdout' do
          it 'should default to an offset of zero' do
            job = Job.new
            state_directory = job.send(:state_directory)

            filehandle = mock('filehandle')
            File.expects(:new).with("#{state_directory}/stdout", 'rb').returns(filehandle)
            filehandle.expects(:seek).with(0, IO::SEEK_SET)
            filehandle.expects(:read).returns("some data")
            filehandle.expects(:close)
            job.stdout.should == "some data"
          end

          it 'should read from an offset' do
            job = Job.new
            state_directory = job.send(:state_directory)

            filehandle = mock('filehandle')
            File.expects(:new).with("#{state_directory}/stdout", 'rb').returns(filehandle)
            filehandle.expects(:seek).with(5, IO::SEEK_SET)
            filehandle.expects(:read).returns("some data at an offset")
            filehandle.expects(:close)
            job.stdout(5).should == "some data at an offset"
          end
        end

        describe '#stderr' do
          it 'should default to an offset of zero' do
            job = Job.new
            state_directory = job.send(:state_directory)

            filehandle = mock('filehandle')
            File.expects(:new).with("#{state_directory}/stderr", 'rb').returns(filehandle)
            filehandle.expects(:seek).with(0, IO::SEEK_SET)
            filehandle.expects(:read).returns("some data")
            filehandle.expects(:close)
            job.stderr.should == "some data"
          end

          it 'should read from an offset' do
            job = Job.new
            state_directory = job.send(:state_directory)

            filehandle = mock('filehandle')
            File.expects(:new).with("#{state_directory}/stderr", 'rb').returns(filehandle)
            filehandle.expects(:seek).with(5, IO::SEEK_SET)
            filehandle.expects(:read).returns("some data at an offset")
            filehandle.expects(:close)
            job.stderr(5).should == "some data at an offset"
          end
        end

        describe '#status' do
          let(:job) do
            test_job = Job.new
            test_job.stubs(:state_directory).returns('test')
            test_job
          end

          it 'should be :failed if there is an error file' do
            File.expects(:exists?).with('test/error').returns(true)
            job.status.should == :failed
          end

          it 'should be :starting if there is no pid' do
            File.expects(:exists?).with('test/error').returns(false)
            File.expects(:exists?).with('test/pid').returns(false)
            job.status.should == :starting
          end

          it 'should be :stopped is there is an exitstatus' do
            File.expects(:exists?).with('test/error').returns(false)
            File.expects(:exists?).with('test/pid').returns(true)
            File.expects(:exists?).with('test/exitstatus').returns(true)
            job.status.should == :stopped
          end

          it 'should be :running if there is no exitstatus' do
            File.expects(:exists?).with('test/error').returns(false)
            File.expects(:exists?).with('test/pid').returns(true)
            File.expects(:exists?).with('test/exitstatus').returns(false)
            job.status.should == :running
          end
        end

        describe '#kill' do
          it 'should send TERM to the pid' do
            job = Job.new
            job.stubs(:pid).returns(58)
            Process.expects(:kill).with('TERM', 58)
            job.kill
          end
        end

        describe '#wait_for_process' do
          it 'should not loop if process is not :running' do
            job = Job.new
            job.expects(:status).returns(:stopped)
            job.expects(:sleep).never
            job.wait_for_process
          end

          it 'should sleep and loop if status is :running' do
            job = Job.new
            job.stubs(:status).returns(:running, :stopped)
            job.expects(:sleep).with(0.1).once
            job.wait_for_process
          end
        end

        describe '#cleanup_state' do
          it 'should blow away the state directory' do
            job = Job.new
            job.stubs(:state_directory).returns('test')
            FileUtils.expects(:remove_entry_secure).with('test')
            job.cleanup_state

            # explicitly unstub so the after block can fire
            FileUtils.unstub(:remove_entry_secure)
          end
        end
      end
    end
  end
end


lib/mcollective/application/shell.rb
require 'mcollective/application/shell/watcher'

class MCollective::Application::Shell < MCollective::Application
  description 'Run shell commands'

  usage <<-END_OF_USAGE
mco shell [OPTIONS] [FILTERS] <ACTION> [ARGS]

  mco shell run [--tail] [COMMAND]
  mco shell start [COMMAND]
  mco shell watch [HANDLE]
  mco shell list
  mco shell kill [HANDLE]
END_OF_USAGE

  option :tail,
         :arguments   => [ '--tail' ],
         :description => 'Switch run to tail mode',
         :type        => :bool

  def post_option_parser(configuration)
    if ARGV.size < 1
      raise "Please specify an action"
    end

    valid_actions = ['run', 'start', 'watch', 'list', 'kill' ]
    action = ARGV.shift

    unless valid_actions.include?(action)
      raise 'Action has to be one of ' + valid_actions.join(', ')
    end

    configuration[:command] = action
  end

  def main
    send("#{configuration[:command]}_command")
  end

  private

  def run_command
    command = ARGV.join(' ')

    if configuration[:tail]
      tail(command)
    else
      do_run(command)
    end
  end

  def start_command
    command = ARGV.join(' ')
    client = rpcclient('shell')

    responses = client.start(:command => command)
    responses.sort_by! { |r| r[:sender] }

    responses.each do |response|
      if response[:statuscode] == 0
        puts "#{response[:sender]}: #{response[:data][:handle]}"
      else
        puts "#{response[:sender]}: ERROR: #{response.inspect}"
      end
    end
    printrpcstats :summarize => true, :caption => "Started command: #{command}"
  end

  def list_command
    client = rpcclient('shell')

    responses = client.list
    responses.sort_by! { |r| r[:sender] }

    responses.each do |response|
      if response[:statuscode] == 0
        next if response[:data][:jobs].empty?
        puts "#{response[:sender]}:"
        response[:data][:jobs].keys.sort.each do |handle|
          puts "    #{handle}"

          if client.verbose
            puts "        command: #{response[:data][:jobs][handle][:command]}"
            puts "        status:  #{response[:data][:jobs][handle][:status]}"
            puts ""
          end
        end
      end
    end

    printrpcstats :summarize => true, :caption => "Command list"
  end

  def watch_command
    handles = ARGV
    client = rpcclient('shell')

    watchers = []
    client.list.each do |response|
      next if response[:statuscode] != 0
      response[:data][:jobs].keys.each do |handle|
        if handles.include?(handle)
          watchers << Watcher.new(response[:sender], handle)
        end
      end
    end

    watch_these(client, watchers)
  end

  def kill_command
    handle = ARGV.shift
    client = rpcclient('shell')

    client.kill(:handle => handle)

    printrpcstats :summarize => true, :caption => "Command list"
  end

  def do_run(command)
    client = rpcclient('shell')

    responses = client.run(:command => command)
    responses.sort_by! { |r| r[:sender] }

    responses.each do |response|
      if response[:statuscode] == 0
        puts "#{response[:sender]}:"
        puts response[:data][:stdout]
        if response[:data][:stderr].size > 0
          puts "    STDERR:"
          puts response[:data][:stderr]
        end
        if response[:data][:exitcode] != 0
          puts "exitcode: #{response[:data][:exitcode]}"
        end
        puts ""
      else
        puts "#{response[:sender]}: ERROR: #{response.inspect}"
      end
    end

    printrpcstats :summarize => true, :caption => "Ran command: #{command}"
  end

  def tail(command)
    client = rpcclient('shell')

    processes = []
    client.start(:command => command).each do |response|
      next unless response[:statuscode] == 0
      processes << Watcher.new(response[:sender], response[:data][:handle])
    end

    watch_these(client, processes, true)
  end

  def watch_these(client, processes, kill_on_interrupt = false)
    client.progress = false

    state = :running
    if kill_on_interrupt
      # trap sigint so we can send a kill to the commands we're watching
      trap('SIGINT') do
        puts "Attempting to stop cleanly, interrupt again to kill"
        state = :stopping

        # if we're double-tapped, just quit (may leave a mess)
        trap('SIGINT') do
          puts "OK you meant it; bye"
          exit 1
        end
      end
    else
      # When we get a sigint we should just exit
      trap('SIGINT') do
        puts ""
        exit 1
      end
    end

    while !processes.empty?
      processes.each do |process|
        client.filter["identity"].clear
        client.identity_filter process.node

        if state == :stopping && kill_on_interrupt
          puts "Sending kill to #{process.node} #{process.handle}"
          client.kill(:handle => process.handle)
        end

        client.status({
          :handle => process.handle,
          :stdout_offset => process.stdout_offset,
          :stderr_offset => process.stderr_offset,
        }).each do |response|
          if response[:statuscode] != 0
            process.flush
            processes.delete(process)
            break
          end

          process.status(response)

          if response[:data][:status] == :stopped
            process.flush
            processes.delete(process)
          end
        end
      end
    end
  end
end


lib/mcollective/application/shell/watcher.rb
require 'mcollective/application/shell/prefix_stream_buf'

# The Watcher class is a utility class for Application::Shell#watch_these.
# It's effectively a tuple of [node, handle] to identify the command, and
# PrefixStreamBufs and watermarks to track where in the stdout/stderr we have
# seen.

module MCollective
  class Application
    class Shell < Application
      class Watcher
        attr_reader :node, :handle
        attr_reader :stdout_offset, :stderr_offset

        def initialize(node, handle)
          @node = node
          @handle = handle
          @stdout = PrefixStreamBuf.new("#{node} stdout: ")
          @stderr = PrefixStreamBuf.new("#{node} stderr: ")
          @stdout_offset = 0
          @stderr_offset = 0
        end

        def status(response)
          @stdout_offset += response[:data][:stdout].size
          @stdout.display(response[:data][:stdout])
          @stderr_offset += response[:data][:stderr].size
          @stderr.display(response[:data][:stderr])
        end

        def flush
          @stdout.flush
          @stderr.flush
        end
      end
    end
  end
end


lib/mcollective/application/shell/prefix_stream_buf.rb
# PrefixStreamBuf is a utility class used my
# MCollective::Application::Shell::Watcher.

# PrefixStreambuf#display takes chunks of input, and on complete lines will
# emit that line with the prefix.  Incomplete lines are kept internally for
# the next call to display, unless #flush is called to flush the buffer.

module MCollective
  class Application
    class Shell < Application
      class PrefixStreamBuf
        def initialize(prefix)
          @buffer = ''
          @prefix = prefix
        end

        def display(data)
          @buffer += data
          chunks = @buffer.lines.to_a
          return if chunks.empty?

          if chunks[-1][-1] != "\n"
            @buffer = chunks[-1]
            chunks.pop
          else
            @buffer = ''
          end

          chunks.each do |chunk|
            puts "#{@prefix}#{chunk}"
          end
        end

        def flush
          if @buffer.size > 0
            display("\n")
          end
        end
      end
    end
  end
end


lib/mcollective/agent/shell/job.rb
require 'securerandom'

# The Job class manages the spawning and state tracking for a process as it's
# running.
#
# The general approach we take is to create a directory per Job
# (Job#state_directory), and populate it as follows:

#   command    - the command we've been asked to run
#   wrapper    - a short program that spawns the command and collects its exit
#                status
#   error      - any problem in spawning the process
#   pid        - the pid of the spawned process
#   stdout     - stdout from the process
#   stderr     - stderr from the process
#   exitstatus - the exitstatus of the spawned process

# The wrapper process is detached from the mcollectived process, so restarting
# the mcollectived will not terminate the process.

module MCollective
  module Agent
    class Shell<RPC::Agent
      class Job
        @@ruby = nil

        # get all the jobs with state
        def self.list
          jobs = []
          Dir.entries(state_path).each do |handle|
            next if handle[0] == '.'
            jobs << Job.new(handle)
          end
          return jobs
        end

        attr_reader :command, :handle, :exitcode, :signal, :pid
        def initialize(handle = nil)
          if !handle
            @handle = SecureRandom.uuid
            Pathname.new(state_directory).mkpath
          else
            @handle = handle

            commandfile = "#{state_directory}/command"
            if File.exists?(commandfile)
              @command = IO.read(commandfile).chomp
            end

            pidfile = "#{state_directory}/pid"
            if File.exists?(pidfile)
              @pid = Integer(IO.read(pidfile))
            end

            get_exitcode
          end
        end

        def start_command(command)
          @command = command
          # We create a manager process which then spawns the command we want
          # to run, the manager then writes a 'pid' file on succesful spawn, or
          # an 'error' file.  Assuming success it waits for completion of the
          # process and records the exit status to the 'exitstatus' file.
          File.open("#{state_directory}/command", 'w') do |fh|
            fh.puts command
          end

          File.open("#{state_directory}/wrapper", 'w') do |fh|
            fh.puts wrapper
          end

          manager = ::Process.spawn(find_ruby, "#{state_directory}/wrapper", {
            :chdir => '/',
            :in => :close,
            :out => :close,
            :err => :close,
          })

          if manager == nil
            raise "Couldn't spawn manager process"
          end

          # busy wait for a pid or error file
          while !File.exists?("#{state_directory}/pid") && !File.exists?("#{state_directory}/error")
            sleep 0.1
          end

          ::Process.detach(manager)

          if File.exists?("#{state_directory}/pid")
            @pid = Integer(IO.read("#{state_directory}/pid"))
          else
            # we must have an error file
            raise IO.read("#{state_directory}/error")
          end
        end

        def stdout(offset = 0)
          fh = File.new("#{state_directory}/stdout", 'rb')
          fh.seek(offset, IO::SEEK_SET)
          out = fh.read
          fh.close
          return out
        end

        def stderr(offset = 0)
          fh = File.new("#{state_directory}/stderr", 'rb')
          fh.seek(offset, IO::SEEK_SET)
          err = fh.read
          fh.close
          return err
        end

        def status
          if File.exists?("#{state_directory}/error")
            # Process failed to start
            return :failed
          end

          if !File.exists?("#{state_directory}/pid")
            # We haven't started yet
            return :starting
          end

          if File.exists?("#{state_directory}/exitstatus")
            # The manager has written out the exitstatus, so the process is done
            return :stopped
          end

          return :running
        end

        def kill
          ::Process.kill('TERM', pid)
        end

        def wait_for_process
          while status == :running
            sleep 0.1
          end
          get_exitcode
        end

        def cleanup_state
          FileUtils.remove_entry_secure state_directory
        end

        private

        def self.state_path
          if Util.windows?
            default = "C:/ProgramData/mcollective-shell"
          else
            default = "/var/run/mcollective-shell"
          end

          Config.instance.pluginconf['shell.state_path'] || default
        end

        def get_exitcode
          statusfile = "#{state_directory}/exitstatus"
          if File.exists?(statusfile)
            status = Integer(IO.read(statusfile))
            @signal = status & 0xff
            @exitcode = status >> 8
          end
        end

        def find_ruby
          if @@ruby
            return @@ruby
          end

          ruby_config = begin
            ::RbConfig::CONFIG
          rescue NameError
            ::Config::CONFIG
          end

          candidates = [
            File.join(ruby_config['bindir'], ruby_config['ruby_install_name']) + ruby_config['EXEEXT'],
            'ruby',
          ]

          found = candidates.find { |path| system('%s -e 42' % path) }

          if found
            @@ruby = found
            return @@ruby
          else
            raise "No ruby found via Config or PATH"
          end
        end

        def wrapper
          return <<-WRAPPER
            command = IO.read("#{state_directory}/command").chomp

            options = {
              :chdir => '/',
              :out => "#{state_directory}/stdout",
              :err => "#{state_directory}/stderr",
            }

            begin
              pid = ::Process.spawn(command, options)

              File.open("#{state_directory}/pid", 'w') do |fh|
                fh.puts pid
              end

              ::Process.waitpid(pid)

              if $?.nil?
                # On win32 $? doesn't seem to get set - probably need to grab a
                # handle then call GetExitCode
                exitstatus = 0
              else
                exitstatus = $?.to_i
              end

              File.open("#{state_directory}/exitstatus", 'w') do |fh|
                fh.puts exitstatus
              end

            rescue Exception => e
              File.open("#{state_directory}/error", 'w') do |fh|
                fh.puts e
              end
            end
          WRAPPER
        end

        def state_directory
          "#{self.class.state_path}/#{@handle}"
        end
      end
    end
  end
end
